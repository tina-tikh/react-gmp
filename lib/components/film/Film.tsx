import * as React from 'react';
import { Component, ReactNode } from 'react';
import { connect } from 'react-redux';

import { ActionBar, ActionBarCaption, Header, Main } from '../layout';
import { Movie } from '../../api';
import { AppState, getSelectedMovie, getSimilarMovies } from '../../store';
import FilmList from '../FilmList';
import FilmDetails from './FilmDetails';
import { LinkButton } from '../form';

interface FilmProps {
  similar: Movie[],
  similarTotal: number,
  movie: Movie,
}

type FilmState = {
  movieId: string
};

class Film extends Component<FilmProps, FilmState> {
  render(): ReactNode {
    return (
      <React.Fragment>
        <Header>
          <LinkButton onClick={() => this.handleBackToSearch()}>Back to search</LinkButton>
          <FilmDetails film={this.props.movie}/>
        </Header>
        <Main>
          <ActionBar>
            <ActionBarCaption>
              Films by {this.props.movie?.genres[0]} genre
            </ActionBarCaption>
          </ActionBar>
          <FilmList films={this.props.similar}/>
        </Main>
      </React.Fragment>
    );
  }

  handleBackToSearch(): void {
    history.back();
  }
}

const mapStateToProps = (state: AppState) => ({
  movie: getSelectedMovie(state),
  similar: getSimilarMovies(state),
  similarTotal: state.selectedMovie.similar.total
});

export default connect(
  mapStateToProps
)(Film);
