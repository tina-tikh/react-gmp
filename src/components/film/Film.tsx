import * as React from 'react';
import { Component, ReactNode } from 'react';
import { match } from 'react-router-dom';
import { connect } from 'react-redux';

import { ActionBar, ActionBarCaption, Header, Main } from '../layout';
import FilmList from '../FilmList';
import { Movie, Movies } from '../../store/types';
import { MovieService } from '../../api';
import { thunkReceiveSimilarByGenre } from '../../thunks';
import { AppState } from '../../store';
import FilmDetails from './FilmDetails';

type FilmProps = {
  match: match<{ id: string }>;
  movies: Movies,
  thunkReceiveSimilarByGenre: typeof thunkReceiveSimilarByGenre
};

type FilmState = {
  movie: Movie;
  genre: string;
};

class Film extends Component<FilmProps, FilmState> {
  componentDidMount(): void {
    this.refresh();
  }

  componentDidUpdate(prevProps: Readonly<FilmProps>): void {
    const pathId = prevProps.match.params.id;
    const prevPathId = this.props.match.params.id;
    if (pathId !== prevPathId) {
      this.refresh();
    }
  }

  refresh() {
    const pathId = this.props.match.params.id;
    MovieService.getMovie(pathId)
      .then((movie) => {
        const genre: string = movie && movie.genres[0];

        this.props.thunkReceiveSimilarByGenre(genre);

        this.setState({
          movie,
          genre,
        });
      });
  }

  render(): ReactNode {
    return (
      <React.Fragment>
        <Header>
          <FilmDetails film={this.state?.movie}/>
        </Header>
        <Main>
          <ActionBar>
            <ActionBarCaption>
              Films by {this.state?.genre} genre
            </ActionBarCaption>
          </ActionBar>
          <FilmList films={this.props?.movies.data}/>
        </Main>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  movies: state.movies
});

export default connect(
  mapStateToProps,
  { thunkReceiveSimilarByGenre }
)(Film);
