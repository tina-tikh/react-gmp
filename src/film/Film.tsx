import * as React from 'react';
import { Component, ReactNode } from 'react';
import { match } from 'react-router-dom';

import {
  ActionBar,
  ActionBarCaption,
  FilmList,
  Header,
  Main,
  Movie,
} from '../common';
import movies from '../common/api/movies-mock.json';
import FilmDetails from './FilmDetails';

type FilmProps = {
  match: match<{ id: string }>;
};

type FilmState = {
  movie: Movie;
  similar: Movie[];
};

class Film extends Component<FilmProps, FilmState> {
  render(): ReactNode {
    // todo: implement api calls
    const pathId = Number(this.props.match.params.id);
    const film = movies.data.find((movie: Movie) => movie.id === pathId);
    const genre = film.genres[0];
    const similar = movies.data;

    return (
      <React.Fragment>
        <Header>
          <FilmDetails film={film} />
        </Header>
        <Main>
          <ActionBar>
            <ActionBarCaption>Films by {genre} genre</ActionBarCaption>
          </ActionBar>
          <FilmList films={similar} />
        </Main>
      </React.Fragment>
    );
  }
}

export default Film;
