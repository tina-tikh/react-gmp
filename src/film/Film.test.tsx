import * as React from 'react';
import { shallow } from 'enzyme';
import { match } from 'react-router-dom';

import Film from './Film';

describe('<Film />', () => {
  const matchParams: match<{ id: string }> = {
    params: {
      id: '123',
    },
    isExact: false,
    path: '',
    url: '',
  };

  it('should be rendered', () => {
    const component = shallow(<Film match={matchParams} />);
    expect(component).toMatchSnapshot();
  });
});
