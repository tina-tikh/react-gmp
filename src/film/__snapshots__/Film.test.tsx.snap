// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Film /> should be rendered 1`] = `
<Fragment>
  <Header>
    <FilmDetails />
  </Header>
  <Styled(main)>
    <Styled(section)>
      <Styled(span)>
        Films by 
         genre
      </Styled(span)>
    </Styled(section)>
    <FilmList />
  </Styled(main)>
</Fragment>
`;
